fix(isolated-declarations): incorrect error when exported type is locally shadowed by an unexported variable

Dunqing opened the pull request:
* close #12465

`Variable`, `Function`, and `TSModuleDeclaration` can be shadowed by type declaration, so we should check if there is a value reference to decide whether it should be retained.

Others like `Enum` and `Class` don't allow redeclaration, so don't need to check.

Boshen commented:
### Merge activity

* **Jul 23, 10:18 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
fix(isolated-declarations): incorrect error when exported type is locally shadowed by an unexported variable (#12466)

* close #12465

`Variable`, `Function`, and `TSModuleDeclaration` can be shadowed by type declaration, so we should check if there is a value reference to decide whether it should be retained.

Others like `Enum` and `Class` don't allow redeclaration, so don't need to check.
